lib/Optimizer/Dialect/FIROps.cpp

```diff
@@ -0,0 +1,573 @@
+      return mlir::failure();
+    WhereOp::ensureTerminator(*elseRegion, parser.getBuilder(),
+                              result.location);
+  }
+
+  // Parse the optional attribute list.
+  if (parser.parseOptionalAttrDict(result.attributes))
+    return mlir::failure();
+
+  return mlir::success();
+}
+
+mlir::ParseResult isValidCaseAttr(mlir::Attribute attr) {
+  if (attr.dyn_cast_or_null<mlir::UnitAttr>() ||
+      attr.dyn_cast_or_null<ClosedIntervalAttr>() ||
+      attr.dyn_cast_or_null<PointIntervalAttr>() ||
+      attr.dyn_cast_or_null<LowerBoundAttr>() ||
+      attr.dyn_cast_or_null<UpperBoundAttr>())
+    return mlir::success();
+  return mlir::failure();
+}
+
+unsigned getCaseArgumentOffset(llvm::ArrayRef<mlir::Attribute> cases,
+                               unsigned dest) {
+  unsigned o = 0;
+  for (unsigned i = 0; i < dest; ++i) {
+    auto &attr = cases[i];
+    if (!attr.dyn_cast_or_null<mlir::UnitAttr>()) {
+      ++o;
+      if (attr.dyn_cast_or_null<ClosedIntervalAttr>())
+        ++o;
+    }
+  }
+  return o;
+}
+
+mlir::ParseResult parseSelector(mlir::OpAsmParser &parser,
+                                mlir::OperationState &result,
+                                mlir::OpAsmParser::OperandType &selector,
+                                mlir::Type &type) {
+  if (parser.parseOperand(selector) || parser.parseColonType(type) ||
+      parser.resolveOperand(selector, type, result.operands) ||
+      parser.parseLSquare())
+    return mlir::failure();
+  return mlir::success();
+}
+
+/// Generic pretty-printer of a binary operation
+void printBinaryOp(Operation *op, OpAsmPrinter &p) {
+  assert(op->getNumOperands() == 2 && "binary op must have two operands");
+  assert(op->getNumResults() == 1 && "binary op must have one result");
+
+  p << op->getName() << ' ' << op->getOperand(0) << ", " << op->getOperand(1);
+  p.printOptionalAttrDict(op->getAttrs());
+  p << " : " << op->getResult(0).getType();
+}
+
+/// Generic pretty-printer of an unary operation
+void printUnaryOp(Operation *op, OpAsmPrinter &p) {
+  assert(op->getNumOperands() == 1 && "unary op must have one operand");
+  assert(op->getNumResults() == 1 && "unary op must have one result");
+
+  p << op->getName() << ' ' << op->getOperand(0);
+  p.printOptionalAttrDict(op->getAttrs());
+  p << " : " << op->getResult(0).getType();
+}
+
+bool isReferenceLike(mlir::Type type) {
+  return type.isa<fir::ReferenceType>() || type.isa<fir::HeapType>() ||
+         type.isa<fir::PointerType>();
+}
+
+mlir::FuncOp createFuncOp(mlir::Location loc, mlir::ModuleOp module,
+                          StringRef name, mlir::FunctionType type,
+                          llvm::ArrayRef<mlir::NamedAttribute> attrs) {
+  if (auto f = module.lookupSymbol<mlir::FuncOp>(name))
+    return f;
+  mlir::OpBuilder modBuilder(module.getBodyRegion());
+  return modBuilder.create<mlir::FuncOp>(loc, name, type, attrs);
+}
+
+GlobalOp createGlobalOp(mlir::Location loc, mlir::ModuleOp module,
```

Comment:
LLVM style is to explicitly qualify functions:
https://llvm.org/docs/CodingStandards.html#use-namespace-qualifiers-to-implement-previously-declared-functions

LLVM style is also to have `using namespace`, and to limit anonymous namespaces to just classes:
https://llvm.org/docs/CodingStandards.html#anonymous-namespaces

Reply:
Ongoing activity.